stan/math/prim/meta/holder.hpp

```diff
@@ -0,0 +1,274 @@
+  EIGEN_STRONG_INLINE void writePacket(Index row, Index col,
+                                       const PacketType& x) {
+    return m_argImpl.template writePacket(row, col, x);
+  }
+  template <int StoreMode, typename PacketType>
+  EIGEN_STRONG_INLINE void writePacket(Index index, const PacketType& x) {
+    return m_argImpl.template writePacket(index, x);
+  }
+};
+
+}  // namespace internal
+}  // namespace Eigen
+
+namespace stan {
+namespace math {
+
+/**
+ * Constructs a no-op operation that also holds pointer to some other
+ * expressions, allocated on heap. When the object is destructed those
+ * expressions will be deleted.
+ * @tparam T type of argument expression
+ * @tparam Ptrs types of pointers
+ * @param a argument expression
+ * @param ptrs pointers to objects the constructed object will own.
+ * @return holder operation
+ */
+template <typename T, typename... Ptrs,
+          std::enable_if_t<sizeof...(Ptrs) >= 1>* = nullptr>
+Holder<T, Ptrs...> holder(T&& arg, Ptrs*... pointers) {
+  return Holder<T, Ptrs...>(std::forward<T>(arg), pointers...);
+}
+// trivial case with no pointers constructs no holder object
+template <typename T>
+T holder(T&& arg) {
+  return std::forward<T>(arg);
+}
+
+namespace internal {
+// the function holder_handle_element is also used in holder_cl
+/**
+ * Handles single element (moving rvalues to heap) for construction of
+ * `holder` or `holder_cl` from a functor. For lvalues just sets the `res`
+ * pointer.
+ * @tparam T type of the element
+ * @param a element to handle
+ * @param res resulting pointer to element
+ * @return tuple of pointer allocated on heap (empty).
+ */
+template <typename T>
+auto holder_handle_element(T& a, T*& res) {
+  res = &a;
+  return std::make_tuple();
+}
+
+/**
+ * Handles single element (moving rvalues to heap) for construction of
+ * `holder` or `holder_cl` from a functor. Rvalue is moved to heap and the
+ * pointer to heap memory is assigned to res and returned in a tuple.
+ * @tparam T type of the element
+ * @param a element to handle
+ * @param res resulting pointer to element
+ * @return tuple of pointer allocated on heap (containing single pointer).
+ */
+template <typename T>
+auto holder_handle_element(std::remove_reference_t<T>&& a, T*& res) {
+  res = new T(std::move(a));
+  return std::make_tuple(res);
+}
+
+/**
+ * Second step in implementation of construction `holder` from a functor.
+ * @tparam T type of the result expression
+ * @tparam Is index sequence for `ptrs`
+ * @tparam Args types of pointes to heap
+ * @param expr result expression
+ * @param ptrs pointers to heap that need to be released when the expression is
+ * destructed
+ * @return `holder` referencing given expression
+ */
+template <typename T, std::size_t... Is, typename... Args>
+auto make_holder_impl_step2(T&& expr, std::index_sequence<Is...>,
+                            const std::tuple<Args*...>& ptrs) {
+  return holder(std::forward<T>(expr), std::get<Is>(ptrs)...);
+}
```

Comment:
Is there a standard place to put boilerplate stuff like index_sequence? I always get confused where to put them and do either the front or end

Reply:
No idea.

Reply:
I guess at the end makes most sense.